syntax = "proto3";

package rpc_info.proto;

// Mensaje para una criptomoneda
message Crypto {
  string id = 1;
  string symbol = 2;
  string name = 3;
  string image = 4;
  double current_price = 5;
  int64 market_cap = 6;
  int32 market_cap_rank = 7;
  int64 fully_diluted_valuation = 8;
  int64 total_volume = 9;
  double high_24h = 10;
  double low_24h = 11;
  double price_change_24h = 12;
  double price_change_percentage_24h = 13;
  string last_updated = 14;
}

// Un punto de precio individual en el tiempo
message HistoricalPricePoint {
  int64 timestamp = 1; // Unix timestamp
  double price = 2;    // Precio en la divisa solicitada
}

// Solicitud para lista de criptos
message CryptoRequest {
  string currency = 1;
  int32 quantity = 2;
}

// Solicitud de historial de precios
message HistoricalRequest {
  string id = 1;          // ID de la criptomoneda (Ej: bitcoin)
  string currency = 2;    // Divisa deseada (Ej: USD, EUR)
  int32 history_size = 3; // Cantidad de puntos históricos a retornar
}

// Respuesta con lista de precios históricos
message HistoricalResponse {
  string id = 1;
  repeated HistoricalPricePoint prices = 2;
}

// Respuesta con lista de criptos
message CryptoList { repeated Crypto cryptos = 1; }

// Solicitud para una cripto específica
message CryptoByIdRequest {
  string id = 1;       // Ej: bitcoin
  string currency = 2; // Ej: USD
}

// Cambio de divisas
message ExchangeRate {
  string currency = 1;
  double rate = 2;
}

message ExchangeRates { repeated ExchangeRate rates = 1; }

message Empty {}

// Servicio gRPC
service CryptoService {
  rpc GetTopCryptos(CryptoRequest) returns (CryptoList);
  rpc GetCryptoById(CryptoByIdRequest) returns (Crypto);

  // RPC para obtener el historial (lectura del caché)
  rpc GetPriceHistory(HistoricalRequest) returns (HistoricalResponse);

  // Stream de datos, para que el cliente no haga polling
  rpc StreamTopCryptos(CryptoRequest) returns (stream CryptoList);

  // Cambio de divisas
  rpc GetExchangeRates(Empty) returns (ExchangeRates);
}